compositor_render/src/renderer.rs

```diff
@@ -42,42 +69,80 @@ pub enum RendererRenderError {
+            shader_transforms: &self.shader_transforms,
+            web_renderers: &self.web_renderers,
+        }
+    }
+
+    pub fn register_transformation(
         &mut self,
-        provider: fn(Rc<WgpuCtx>) -> T,
+        key: TransformationRegistryKey,
+        spec: TransformationSpec,
     ) -> Result<(), RendererRegisterTransformationError> {
-        self.registry
-            .register(Box::new(provider(self.wgpu_ctx.clone())))?;
-
+        match spec {
+            TransformationSpec::Shader { source } => self
+                .shader_transforms
+                .register(&key, Arc::new(Shader::new(&self.ctx(), source)))?,
+            TransformationSpec::WebRenderer(params) => self
+                .web_renderers
+                .register(&key, Arc::new(WebRenderer::new(&self.ctx(), params)?))?,
+        };
         Ok(())
     }
 
     /// This is very much a work in progress.
     /// For now it just takes a random frame from the input and returns it
-    pub fn render(&self, inputs: InputFrames) -> Result<Frame, RendererRenderError> {
-        let output_frame = inputs.frames.values().next().cloned();
-
-        match output_frame {
-            Some(frame) => {
-                let frame = Frame {
-                    data: frame.data.clone(),
-                    resolution: frame.resolution,
-                    pts: inputs.pts,
-                };
-                Ok(frame)
-            }
-            None => Err(RendererRenderError::NoInput(0)),
+    pub fn render(
+        &self,
+        mut inputs: FrameSet<InputId>,
+    ) -> Result<FrameSet<OutputId>, RendererRenderError> {
+        let pts = inputs.frames.iter().next().unwrap().1.pts;
+        let ctx = self.ctx();
+        populate_inputs(&ctx, &self.scene, &mut inputs.frames);
+        run_transforms(&ctx, &self.scene);
+        let mut result = HashMap::new();
+        for (node_id, output) in &self.scene.outputs {
+            // TODO: very temporary implementation
+            // TODO: convert rgba to yuv
+            let yuv_data = output.1.download(&self.wgpu_ctx);
+            result.insert(
+                node_id.clone(),
+                Arc::new(Frame {
+                    data: yuv_data,
+                    resolution: output.1.resolution,
+                    pts,
+                }),
+            );
         }
+        error!("TODO: convert rgba back to yuv and read into memory");
+        Ok(FrameSet {
+            frames: result,
+            pts: inputs.pts,
+        })
     }
 
-    pub fn update_scene(&mut self, scene: Scene) {
-        self.scene = Some(scene);
+    pub fn update_scene(&mut self, scene_specs: SceneSpec) -> Result<(), SceneUpdateError> {
+        self.scene.update(
+            &RenderCtx {
+                wgpu_ctx: &self.wgpu_ctx,
+                electron: &self.electron_instance,
+                shader_transforms: &self.shader_transforms,
+                web_renderers: &self.web_renderers,
+            },
```

Comment:
I'm borrowing scene mutably there too, if I use ctx function rust will consider entire self to be borrowed and it will not allow that code, but if I use literal compiler can see that I can both borrow those fields and mut scene

Reply:
oh, true. thst's sad, I thought it would understand it